### PR TITLE
Issue/#2103 alt - define coding standards for Friendica

### LINKED ISSUE #2103: define a coding style standard for friendica
Many parts of friendica does use different coding styles. This does make reading and understanding the code harder than it should be. In addition it's hard to search for a certain piece of code because you have to know all the possible different coding styles. 
I would unify some code but at the present time it is unclear what style is the standard one.

To make developing easier and to attract more developers we should agree on a standard coding style for friendica.
To give a simple example: https://hub.hfrc.de/help/developers&JS=1

This issue should stay open to collect the topics/items/pieces which should go into the standard coding style documentation.
To keep this page readable **please use the friendica forums for the discussion** about which coding style is better or which is not.

Things which come into my mind:
- SQL queries
- `$array['key']` or `$array["key"]`
- deprecated functions:
  - using `z_root()` instead of `$a->get_baseurl()`

### PR DESCRIPTION
**Pull request by Hypolite**
This lightweight PR just adds the PHP CS Fixer tool to the util folder and update the documentation related to coding standards.

Fixes #2103, code formatting updates will happen over time.

### CODE REVIEW

File: doc/Developers-Intro.md [hunk 1]
```diff
@@ -42,29 +42,43 @@ If you have seen Friendica you probably have ideas to improve it, haven't you?
...
 We can't promise we have the right skills in the group but we'll try.
-* Choose a thing to start with, e.g. work on the icon set of your favourite theme
+* Choose a thing to start with, e.g. work on the icon set of your favorite theme
 
 Programming
 ---
 
+###Coding standards
+
+For the sake of consistency between contribution and general code readability, Friendica follows the widespread [PSR-2 coding standards](http://www.php-fig.org/psr/psr-2/). Don't worry, you don't have to know by heart the PSR-2 coding standards. Before comitting your work, you can use the PHP Coding Standard Fixer located in the `util` folder this way:
```
Comment (tobiasd) [inline doc/Developers-Intro.md]: We are using the rule of one-sentence-per-line for the docs. Could you change this.

Comment (fabrixxm) [inline doc/Developers-Intro.md]: a coding standard for documentation about coding standards :-P

Comment (Hypolite) [inline doc/Developers-Intro.md]: @tobiasd Makes sense, I just wish I knew/guessed it before.

@fabrixxm: Ha ha!

File: doc/Developers-Intro.md [hunk 2]
```diff
@@ -42,29 +42,45 @@ If you have seen Friendica you probably have ideas to improve it, haven't you?
...
 Programming
 ---
 
+###Coding standards
+
+For the sake of consistency between contribution and general code readability, Friendica follows the widespread [PSR-2 coding standards](http://www.php-fig.org/psr/psr-2/).
+Don't worry, you don't have to know by heart the PSR-2 coding standards to start contributing to Friendica.
+Before comitting your work, you can use the PHP Coding Standard Fixer located in the `util` folder this way:
+
+	$> php util/php-cs-fixer.phar fix <file or folder> --level=psr2
```
Comment (Hypolite) [inline doc/Developers-Intro.md]: For what it's worth, this tool is standalone, it just needs PHP to be executed, which is required for Friendica anyway, but I totally get your point. I'll do some rewriting.

**Comment by Hypolite** [comment]
Updated the documentation with the latest decision to keep on with tabs.

**Comment by Hypolite** [comment]
Can this be merged sooner than later? We already are enforcing it in the current PRs.

**Comment by annando** [comment]
I think everybody of the regular developers should agree on the rules before officially accepting it. And I think the rules shouldn't be manifested. In the next time we should apply small changes to the rules.

**Comment by rabuzarus** [comment]
> In the next time we should apply small changes to the rules.

But I would be nice if this future changes wouldn't influence automated code formatting tools. Because this would one of the big improvements of having an official widely used coding standard. My IDE can do the formatting for me :-)

**Comment by annando** [comment]
Mine not :-)

**Comment by Hypolite** [comment]
`nano` really cannot be called an IDE, hence the missing features. What small changes would you like to apply to the rules?

@rabuzarus: PHPCodeSniffer can be tweaked, and we can publish the exact config we use according to the rules we chose.

**Comment by strk** [comment]
See http://editorconfig.org/ as a way to encode some of the requested style in an editor-agnostic config file (vim supports it too)

**Comment by annando** [comment]
I have mainly three red lines that I personally wouldn't like to cross concerning the code styles:

1. Forced use of ```||``` and ```&&``` in conditions
2. Spaces instead of tabs for indention
3. Linefeed between function definition and brace.

(Mostly) everything else is no problem (like using braces even with oneliners or use of ```'``` on most places) or is definitely something I already wanted to change myself like spaces between ```if``` and the following brace or spaces between mathematical and most logical operators.

You are free to call me eccentric concerning these points :-)

So I would like that we follow the PSR rules according to the spirit and not to the words. The goal should be code that is easy maintainable and understandable not only for PHP professionals but for most average spare time coders.

**Comment by rabuzarus** [comment]
> So I would like that we follow the PSR rules according to the spirit and not to the words. The goal should be code that is easy maintainable and understandable not only for PHP professionals but for most average spare time coders.

Yes and that's the reason for coding standards. So you wouldn't have to re-learn how to read and write php code when you switch between PHP projects.

**Comment by annando** [comment]
Yeah, but you could still create code that complies with the PSR rules and is hardly readable, I guess. (For example by using several layers of class inheritance, ultra long functions, no code comments at all, (nearly) similar code that is repeated multiple times, ...

**Comment by rabuzarus** [comment]
Yes coding standard doesn't mean that it wouldn't be possible to write non-understandable code. But coding standards are one link in the chain. 
There will be always some things, which aren't solvable  with standards because they depend on the context (e.g. inline comments). But this doesn't mean that we can lower the barriers for inter exchange where we could.

It is also good for newcomers. You will learn something about standards which will help you in other projects as well. You don't need to learn whats the specific way of formatting of the project is. You can write in one project in the morning and in another in the evening without rethinking about formatting issues or to reconfigure the IDE auto-formatting feature. If you have learned it you can focus on writing code. It's like with an advanced editor. The beginning could be much easier. But after a very short time you will see the advantage. And the best thing is you can use this knowledge in in many cases and it will simplify your life.

**Comment by Quix0r** [comment]
Maybe related: #3010

**Comment by Quix0r** [comment]
But please, if such guide-lines are being written, try to keep them short&sweet. Nobody wants to read through tons of Wiki pages only to participate on a project, I guess. :-)

**Comment by Hypolite** [comment]
It's already being written in the form of the PSR-2 code styling guide, although the following practice seemed to emerge: we don't change too much of a file if it's too big because of the difficulty to review so much changes. Instead, we change some formatting only in the functions we're working on (mandatory braces for conditions), and we keep some other formatting consistent with the rest of the file (string quotation, concatenation spaces, AND/&&)

**Comment by Hypolite** [comment]
@Quix0r : Here's my proposed changes to the Docs: https://github.com/friendica/friendica/pull/2841/files

Is it short and sweet enough for you?

**Comment by Quix0r** [comment]
Looks fine with me. :-)

File: doc/Developers-Intro.md [hunk 3]
```diff
@@ -42,29 +42,64 @@ If you have seen Friendica you probably have ideas to improve it, haven't you?
...
+ * Indentation is tabs, period (not PSR-2).
+ * By default, strings are enclosed in single quotes, but feel free to use double quotes if it makes more sense (SQL queries, adding tabs and line feeds).
+ * Operators are wrapped by spaces, e.g. `$var === true`, `$var = 1 + 2` and `'string' . $concat . 'enation'`
+ * Braces are mandatory in conditions
+ * No closing PHP tag
+ * No trailing spaces
+
+Don't worry, you don't have to know by heart the PSR-2 coding standards to start contributing to Friendica.
+There are a few tools you can use to check or fix your files before you commit.
+
```
Comment (tobiasd) [inline doc/Developers-Intro.md]: One thing that crossed my mind--even though not strictly coding standard--could you add the "one line per sentence" rule for the docs below the actual coding standards?

**Comment by tobiasd** [comment]
> PHPCodeSniffer can be tweaked, and we can publish the exact config we use according to the rules we chose.

could you write them @Hypolite?

**Comment by Hypolite** [comment]
@tobiasd Yes, but so far we have been unable to agree on app-wide coding standards. We are currently operating in a heavily context-sensitive mode where the rest of the function or the file is used to infer the standards to use. A CodeSniffer custom config would currently bring up many defaults that are now considered OK because the rest of the file/function is using a different standard.

**Comment by tobiasd** [comment]
Ok, makes sense. Then it's only the line about the one-sentence-per-line coding standard for the docs you wanted to add ;-)

**Pull request merged**